File: rye/src/main/java/com/rysonw/rye/frontend/scanner.java
package com.craftinginterpreters.rysonw.rye;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.rysonwinterpreter.rye.*;

class Scanner {
    private final String source; // Filepaths only
    private final List<Token> tokens = new ArrayList<>();
    private static final Map<String, TokenType> keywords;
    private int start = 0;
    private int current = 0;
    private int line = 1;

    static {
        keywords = new HashMap<>();
        keywords.put("and",    AND);
        keywords.put("class",  CLASS);
        keywords.put("else",   ELSE);
        keywords.put("false",  FALSE);
        keywords.put("for",    FOR);
        keywords.put("fun",    FUN);
        keywords.put("if",     IF);
        keywords.put("nil",    NIL);
        keywords.put("or",     OR);
        keywords.put("print",  PRINT);
        keywords.put("return", RETURN);
        keywords.put("super",  SUPER);
        keywords.put("this",   THIS);
        keywords.put("true",   TRUE);
        keywords.put("var",    VAR);
        keywords.put("while",  WHILE);
    }

    // Source file is one big string
    Scanner(String source) {
        this.source = source;
    }

    List<Token> scanTokens() {
        while (!isAtEnd()) {
            start = current;
            scanToken();
        }

        tokens.add(new Token(EOF, "", null, line));
        return tokens;
    }

    private boolean match(char expected) {
        if (isAtEnd()) {
            return false;
        }

        if (source.charAt(current) != expected) {
            return false;
        }

        current++;
        return true;
    }

    private boolean isAtEnd() {
        return current >= source.length();
    }

    private void scanToken() {
        char c = advance();

        switch (c) {
            case '(': addToken(LEFT_PAREN);  break;
            case ')': addToken(RIGHT_PAREN); break;
            case '{': addToken(LEFT_BRACE);  break;
            case '}': addToken(RIGHT_BRACE); break;
            case ',': addToken(COMMA);       break;
            case '.': addToken(DOT);         break;
            case '-': addToken(MINUS);       break;
            case '+': addToken(PLUS);        break;
            case ';': addToken(SEMICOLON);   break;
            case '*': addToken(STAR);        break;

            // Cases where we need to check the char in front
            case '!':
                addToken(match('=') ? BANG_EQUAL : BANG);
                break;

            case '=':
                addToken(match('=') ? EQUAL_EQUAL : EQUAL);
                break;

            case '<':
                addToken(match('=') ? LESS_EQUAL : LESS);
                break;

            case '>':
                addToken(match('=') ? GREATER_EQUAL : GREATER);
                break;

            // Cases where we need to check a char is not in front (Division OP vs Comment)
            case '/':
                if (match('/')) { // match() looks at next char and consumes it
                    // Comments are consumed and we move to the next token
                    while (peek() != '\n' && !isAtEnd()) {
                        advance();
                    }
                }
                else {
                    addToken(SLASH);
                }
                break;

            case ' ': break;
            case '\r': break;
            case '\t': break;
            case '\n': line++; break;
            case"".string(): break;

            case 'o':
            if (match('r')) {
                addToken(OR);
            }
            break;

            default:
                if (isDigit(c)) {
                    number();
                } 
                else if (isAlpha(c)) {
                    identifier();
                } 
                else {
                    Lox.error(line, "Unexpected character.");
                }
        }
    }

    // Helpers
    private char peek() {
        if (isAtEnd()) return '\0';
        return source.charAt(current);
    }

    private void identifier() {
        while (isAlphaNumeric(peek())) advance();

        String text = source.substring(start, current);
        TokenType type = keywords.get(text);
        if (type == null) type = IDENTIFIER;
        addToken(type);
    }



    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') ||
            (c >= 'A' && c <= 'Z') ||
                c == '_';
    }

    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }

    private void number() {
        while (isDigit(peek())) advance();

        // Look for fraction
        if (peek() == '.' && isDigit(peekNext())) {
        advance();

        while (isDigit(peek())) advance();
        }

        addToken(NUMBER,
            Double.parseDouble(source.substring(start, current)));
    }

    private char peekNext() {
        if (current + 1 >= source.length()) return '\0';
        return source.charAt(current + 1);
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private void string() {
        while (peek() != '"' && !isAtEnd()) {
        if (peek() == '\n') line++;
            advance();
        }

        if (isAtEnd()) {
            Lox.error(line, "Unterminated string.");
            return;
        }

        advance();

        String value = source.substring(start + 1, current - 1);
        addToken(STRING, value);
    }


    private char advance() {
        return source.charAt(current++);
    }

    private void addToken(TokenType type) {
        addToken(type, null);
    }

    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }
}

File: rye/src/main/rye/frontend/Scanner.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace com.rysonw.rye.frontend
{
    public class Scanner
    {
        private readonly string source;
        private readonly List<Token> tokens = new List<Token>();
        private int start = 0;
        private int current = 0;
        private int line = 1;

        public Scanner(string source)
        {
            this.source = source ?? string.Empty;
        }

        public List<Token> ScanTokens()
        {
            while (!IsAtEnd())
            {
                start = current;
                ScanToken();
            }

            tokens.Add(new Token(TokenType.EOF, string.Empty, null, line));
            return tokens;
        }

        private void ScanToken()
        {
            char c = Advance();

            switch (c)
            {
                // Single
                case '(': AddToken(TokenType.LEFT_PAREN); break;
                case ')': AddToken(TokenType.RIGHT_PAREN); break;
                case '{': AddToken(TokenType.LEFT_BRACE); break;
                case '}': AddToken(TokenType.RIGHT_BRACE); break;
                case ',': AddToken(TokenType.COMMA); break;
                case '.': AddToken(TokenType.DOT); break;
                case '-': AddToken(TokenType.MINUS); break;
                case '+': AddToken(TokenType.PLUS); break;
                case ';': AddToken(TokenType.SEMICOLON); break;
                case '*': AddToken(TokenType.STAR); break;

                // Double
                case '!': AddToken(match('=') ? TokenType.BANG_EQUAL : TokenType.BANG); break;
                case '=': AddToken(match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL); break;
                case '<': AddToken(match('=') ? TokenType.LESS_EQUAL : TokenType.LESS); break;
                case '>': AddToken(match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER); break;
                case '/':
                    if (match('/'))
                    {
                        // Find the position for end of comment
                        while (Peek() != '\n' && !IsAtEnd())
                        {
                            Advance();
                        }
                    }
                    else
                    {
                        AddToken(TokenType.SLASH);
                    }
                    break;

                // Whitespace (IGNORE)
                case ' ':
                case '\r':
                case '\t':
                    break;
                case '\n':
                    line++;
                    break;

                // Literals
                case '"': CheckString(); break;
                
                default:

                    if (IsDigit(c))
                    {
                        Number();
                    }
                    else
                    {
                        Rye.Error(line, "Unexpected Character.");
                    }

                    Rye.Error(line, "Unexpected error reading token...");
                    break;
            }
        }

        private bool match(char expected)
        {
            if (IsAtEnd()) { return false; }
            if (source[current] != expected) { return false; }
            current++;
            return true;
        }

        private void AddToken(TokenType type) {
            AddToken(type, null);
        }

        private void AddToken(TokenType type, Object literal) {
            Console.WriteLine($"Start: {start}; Current: {current}");
            String text = source.Substring(start, current - start);
            tokens.Add(new Token(type, text, literal, line));
        }

        private void CheckString()
        {
            while (Peek() != '"' && !IsAtEnd())
            {
                if (Peek() == '\n') line++;
                Advance();
            }

            if (IsAtEnd())
            {
                Rye.Error(line, "Unterminated String");
                return;
            }

            Advance();

            int length = current - start - 2;
            string value = length > 0 ? source.Substring(start + 1, length) : string.Empty;
            AddToken(TokenType.STRING, value);
        }

        private bool IsDigit(char c) {
            return c >= '0' && c <= '9';
        }

        private void Number()
        {
            while (IsDigit(Peek())) Advance();

            if (Peek() == '.' && IsDigit(PeekNext()))
            {
                Advance(); // consume '.'
                while (IsDigit(Peek())) Advance();
            }

            string text = source.Substring(start, current - start);
            if (double.TryParse(text, System.Globalization.NumberStyles.Number,
                                System.Globalization.CultureInfo.InvariantCulture,
                                out double value))
            {
                AddToken(TokenType.NUMBER, value);
            }
            else
            {
                Rye.Error(line, $"Invalid number literal: {text}");
            }
        }

        private bool IsAtEnd() => current >= source.Length;
        private char Advance() => source[current++];
        private char Peek() => IsAtEnd() ? '\0' : source[current]; 
        private char PeekNext() => (current + 1) >= source.Length ? '\0' : source[current + 1];
    }
}
